sky/provision/kubernetes/network.py

```diff
@@ -0,0 +1,217 @@
+from typing import Any, Dict, List, Optional
+
+from sky.adaptors import kubernetes
+from sky.provision import common
+from sky.provision.kubernetes import network_utils
+from sky.utils import kubernetes_enums
+from sky.utils.resources_utils import port_ranges_to_set
+
+_PATH_PREFIX = '/skypilot/{cluster_name_on_cloud}/{port}'
+_LOADBALANCER_SERVICE_NAME = '{cluster_name_on_cloud}-skypilot-loadbalancer'
+
+
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, 'provider_config is required'
+    port_mode = network_utils.get_port_mode(
+        provider_config.get('port_mode', None))
+    ports = list(port_ranges_to_set(ports))
+    if port_mode == kubernetes_enums.KubernetesPortMode.LOADBALANCER:
+        _open_ports_using_loadbalancer(
+            cluster_name_on_cloud=cluster_name_on_cloud,
+            ports=ports,
+            provider_config=provider_config)
+    elif port_mode == kubernetes_enums.KubernetesPortMode.INGRESS:
+        _open_ports_using_ingress(cluster_name_on_cloud=cluster_name_on_cloud,
+                                  ports=ports,
+                                  provider_config=provider_config)
+
+
+def _open_ports_using_loadbalancer(
+    cluster_name_on_cloud: str,
+    ports: List[int],
+    provider_config: Dict[str, Any],
+) -> None:
+    service_name = _LOADBALANCER_SERVICE_NAME.format(
+        cluster_name_on_cloud=cluster_name_on_cloud)
+    content = network_utils.fill_loadbalancer_template(
+        namespace=provider_config['namespace'],
+        service_name=service_name,
+        ports=ports,
+        selector_key='skypilot-cluster',
+        selector_value=cluster_name_on_cloud,
+    )
+    network_utils.create_or_replace_namespaced_service(
+        namespace=provider_config['namespace'],
+        service_name=service_name,
+        service_spec=content['service_spec'])
+
+
+def _open_ports_using_ingress(
+    cluster_name_on_cloud: str,
+    ports: List[int],
+    provider_config: Dict[str, Any],
+) -> None:
+    if not network_utils.ingress_controller_exists():
+        raise Exception(
+            "Ingress controller not found. Please install ingress controller first."
```

Comment:
nit: use single quote for strs (here and elsewhere in code so format.sh passes)

Reply:
Done.